transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,155 @@
+
+_In cloud environments_, when hosting entities (e.g. containers, pods, virtual machines, etc.) are scaled down, their state is erased, e.g. their file system, ip address, and memory are deleted. This is exactly the situation where WildFly cannot guarantee that all transactions will be completed and a data integrity issue might occur.
+
+_1.2 There are additional complications when hosting entities crash (i.e. an unintentional shutdown occurs). This extreme scenario is beyond what WildFly can handle and the hosting cloud environment must provide the ability to recover the hosting entity from crashes._
+
+The state of the hosting entity (e.g. its file system, ip address, etc.) should be recovered by the cloud environment without the intervention of the user. In K8s, for example, the employment of StatefulSet creates a stable environment where crashes are handled automatically. If a particular cloud environment does not provide a guarantee similar to StatefulSet, it is not possible to employ WildFly’s Transaction Subsystem.
+
+*As a solution to solve problem 1.2 cannot be developed in WildFly (as it is impossible to control the lifecycle of the host from WildFly), this proposal is only concerned with problem 1.1.*
+
+== Issue Metadata
+
+=== Issue
+
+* [WFLY-17742](https://issues.jboss.org/browse/WFLY-17742)
+
+=== Dev Contacts
+
+* mailto:{jfinelli@redhat.com}[{Manuel Finelli}]
+
+=== QE Contacts
+
+* mailto:{istraka@redhat.com}[{Ivan Straka}]
+
+=== Testing By
+
+* [ ] Engineering
+* [X] QE
+
+=== Affected Projects or Components
+
+* [wildfly-core](https://github.com/wildfly/wildfly-core)
+* [wildfly](https://github.com/wildfly/wildfly)
+* [narayana](https://github.com/jbosstm/narayana)
+
+=== Relevant Installation Types
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* The Transactions subsystem should become aware of how many in-doubt and heuristic transactions there are in the Object Store and delay its suspension as long as all transactions are completed. This would guarantee data integrity in all scenarios, including situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
+* With regards to the suspension logic of the EJB subsystem ([reference](https://github.com/wildfly/wildfly/blob/main/ejb3/src/main/java/org/jboss/as/ejb3/suspend/EJBSuspendHandlerService.java#L203)), the suspension handler of this subsystem should retain control (i.e. it should not return) as long as there are in-doubt transactions and heuristic transactions to be completed. This would guarantee data integrity, especially in situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
```

Comment:
I'd prefer to keep the discussion open as also other engineers are contributing to it. Thanks for your feedback :-)

Reply:
But the proposal cannot be reviewed properly in its current state because you are awaiting the outcome of other discussions. If you add the `hold` label then that will be an indicator that this one isn't ready for review yet and it's more of a discussion document rather than a proposal.

Reply:
Label added, thanks!